lib/lf/geometry/segment_o2.h

```diff
@@ -0,0 +1,55 @@
+/**
+ * @file
+ * @brief Declaration of second-order segments
+ * @author Anian Ruoss
+ * @date   2018-11-18 19:02:17
+ * @copyright MIT License
+ */
+
+#ifndef SEGMENT_O2_H
+#define SEGMENT_O2_H
+
+#include "geometry_interface.h"
+
+namespace lf::geometry {
+
+/**
+ * @brief A curved edge parametrized by means of polynomial of degree 2 defined
+ * by the location of its two endpoints and its midpoint
+ */
+class SegmentO2 : public Geometry {
+ public:
+  explicit SegmentO2(Eigen::Matrix<double, Eigen::Dynamic, 3> coords)
+      : coords_(std::move(coords)) {}
+
+  dim_t DimLocal() const override { return 1; }
+  dim_t DimGlobal() const override { return coords_.rows(); }
+  base::RefEl RefEl() const override { return base::RefEl::kSegment(); }
+  Eigen::MatrixXd Global(const Eigen::MatrixXd& local) const override;
+  Eigen::MatrixXd Jacobian(const Eigen::MatrixXd& local) const override;
+  Eigen::MatrixXd JacobianInverseGramian(
+      const ::Eigen::MatrixXd& local) const override;
+  Eigen::VectorXd IntegrationElement(
+      const Eigen::MatrixXd& local) const override;
+  std::unique_ptr<Geometry> SubGeometry(dim_t codim, dim_t i) const override;
+
+  /** @brief creation of child geometry for the sake of mesh refinement
+   *
+   * @see Geometry::ChildGeometry()
+   * @see RefinementPattern
+   * @param ref_pat three refinement patterns are supported
+   * - rp_nil: empty refinement
+   * - rp_copy: just copies the geometry information of the segment
+   * - rp_split: split edge in the middle.
+   * @param codim _relative_ codimension of children whose shape is requested
+   */
+  std::vector<std::unique_ptr<Geometry>> ChildGeometry(
+      const RefinementPattern& ref_pat, base::dim_t codim) const override;
+
+ private:
```

Comment:
A polynomial has many different representations corresponding to choosing different bases. For instance, choosing the Lagrangian basis for a set of interpolation nodes renders evaluating the polynomial at these notes very efficient. Conversely, this evaluation entails running the Horner scheme, if one chooses the monomial basis. The task is to find a basis for the space of quadratic polynomials so that the evaluations in `Global()` etc can be performed with a minimal number of evaluations.

Reply:
I think it's important to realize that most 2nd order mesh generators (e.g. GMSH) parametrize 2nd order elements with a Lagrangian basis that places the nodes on the corners and midpoints. So from a practicability point of view I think it's important that the constructor of `SegmentO2` can be initialized with the position of the Lagrangian nodes. But as @hiptmair points out, one can of course represent the second order polynomial internally in other ways.
Horners scheme is a nearly perfect solution because it
- minimizes the number of additions/multiplications
- it is numerically stable

For the implementation this would mean that the coefficients of the horner scheme should be computed in the constructor from the coordinates of the Lagrangian nodes.

Personally I've never implemented second order elements using horners scheme but I think it's a very interesting option.

Reply:
What I forgot to say: Of course, if we know where we are going to evaluate the polynomial in advance we can choose a specialized basis so as to minimize the number of additions/multiplications. But I would say that every decently flexible FEM code cannot make such assumptions.
Therefore I think that the Horner Scheme is a very good option.

Reply:
Thank you for all your explanations. All changes are implemented now :+1: